### PR TITLE
New serverless pattern - EventBridge-Bedrock-S3-AOSS

### PR DESCRIPTION
**Pull request by rajavaid77**
*Issue #, if available:2382

*Description of changes:*
Added a serverless pattern to automate syncing an S3 data source sources to knowledge base for Amazon Bedrock using EventBridge scheduler. An EventBridge schedule is configured to run at a regular rate (default of 5mins). The EventBridge uses the universal target to trigger the StartIngestionJob API on the Amazon Bedrock Agent service. The StartIngestionJob operation in turn would sync the data source so that it is re-indexed to the knowledge base. Syncing is incremental, so Amazon Bedrock only processes added, modified, or deleted documents since the last sync.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by biswanathmukherjee** [comment]
I am getting this error while deploying:

`1:26:17 PM | CREATE_FAILED        | AWS::Logs::Delivery            | BedrockKBDelivery
Resource handler returned message: "Supplied Policy document is breaching Cloudwatch Logs policy length limit. (Service: AWSLogs; Status Code: 400; Error
Code: AccessDeniedException; Request ID: 5f1e6680-e223-4073-b440-033fadb60a0b; Proxy: null)" (RequestToken: e55167e6-3cf6-7017-b968-b5307db4d3e5, Handler
ErrorCode: AccessDenied)
1:26:17 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack     | BedrockKBStack
The following resource(s) failed to create: [BedrockKBDelivery]. Rollback requested by user.
1:26:17 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack     | BedrockKBStack
The following resource(s) failed to create: [BedrockKBDelivery]. Rollback requested by user.`

File: eventbridge-bedrock-s3-aoss/README.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+## How it works
+Upon deployment, the CDK stack will create a Knowledge Base for Bedrock configure with S3 Bucket as data source and an OpenSearch Serverless collection to store vector data. A data source repository contains files or content with information that can be retrieved when your knowledge base is queried. The stack also include an EventBridge scheduler that is configured to run every 5 mins and invoke the `StartIngestionJob` API on Amazon Bedrock Agents service. Amazon Bedrock supports a monitoring system to help you understand the execution of any data ingestion jobs. The Stack would create the neccessary CloudWatch log groups and CloudWatch delivery. You can gain visibility into the ingestion of your knowledge base resources with this logging system. Additionally, Amazon Bedrock is integrated with AWS CloudTrail, a service that provides a record of actions taken by a user, role, or an AWS service in Amazon Bedrock. CloudTrail captures all API calls for Amazon Bedrock as events.
```
Comment (biswanathmukherjee) [inline eventbridge-bedrock-s3-aoss/README.md:113]: Upon deployment, the CDK stack will create a Knowledge Base for Bedrock configured with S3 Bucket as data source and an OpenSearch Serverless collection to store vector data.

Comment (biswanathmukherjee) [reply eventbridge-bedrock-s3-aoss/README.md:113]: Pls use Amazon Bedrock at the beginning and for rest of the file use Bedrock. Same naming standard for all other AWS services.

File: eventbridge-bedrock-s3-aoss/example-pattern.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
+{
+  "title": "# EventBridge to Amazon Bedrock to Amazon OpenSearch Serverless",
```
Comment (biswanathmukherjee) [inline eventbridge-bedrock-s3-aoss/example-pattern.json:2]: Pls give a meaningful pattern name that describes the functionality of this pattern. No need to add `#` at the beginning.

File: eventbridge-bedrock-s3-aoss/example-pattern.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  "description": "Periodically sync Bedrock Knowledge Base",
```
Comment (biswanathmukherjee) [inline eventbridge-bedrock-s3-aoss/example-pattern.json:3]: Pls elaborate the description a bit but under 150 char.

File: eventbridge-bedrock-s3-aoss/example-pattern.json [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  "authors": [
+      {
+          "name": "Raja Vaidyanathan",
+          "image": "https://avatars.githubusercontent.com/u/125882914?v=4",
+          "bio": "Solutions Architect @ AWS",
+          "linkedin": "rajav",
+          "twitter": "rajav"
+      }
+  ],
+  "patternArch": {
```
Comment (biswanathmukherjee) [inline eventbridge-bedrock-s3-aoss/example-pattern.json:62]: Pls remove this section.

**Comment by biswanathmukherjee** [comment]
It seems that the pattern will periodically keep call Bedrock API to initiate KB sync even if there is no change in the source files? If so that will cause unnecessary API invocation costs. Instead of this, can we enhance this pattern to trigger the sync whenever there is a change in any source file and sync only that specific file from the bucket?

**Review by biswanathmukherjee — CHANGES_REQUESTED** [review]

**Comment by rajavaid77** [comment]
> It seems that the pattern will periodically keep call Bedrock API to initiate KB sync even if there is no change in the source files? If so that will cause unnecessary API invocation costs. Instead of this, can we enhance this pattern to trigger the sync whenever there is a change in any source file and sync only that specific file from the bucket?

Amazon Bedrock only processes added, modified, or deleted documents since the last sync. So cost will be incurred only when there are changes to be applied. Also the pattern caters to a use case when customers want to have a scheduled sync so they can sync the KB on a periodic basis rather that for each update happening on the datasource, for example in a daily/weekly batch inference scenario etc. let me know if this makes sense. Thanks

File: eventbridge-bedrock-s3-aoss/README.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
+# EventBridge to Amazon Bedrock to Amazon OpenSearch Serverless
```
[the remark this answers is not in the document]
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/README.md]: updated

File: eventbridge-bedrock-s3-aoss/example-pattern.json (hunk 4 above)
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/example-pattern.json]: removed patternArch section

File: eventbridge-bedrock-s3-aoss/example-pattern.json (hunk 3 above)
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/example-pattern.json]: done

File: eventbridge-bedrock-s3-aoss/example-pattern.json (hunk 2 above)
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/example-pattern.json]: done

File: eventbridge-bedrock-s3-aoss/README.md (hunk 1 above)
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/README.md]: updated service names accordingly and fixed grammar

File: eventbridge-bedrock-s3-aoss/README.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+6. Use the checkboxes to select the models you wish to enable. Review the applicable EULAs as needed. Click **Next** to go to the Review screen and then **Submit** to enable the required models in your account. For this pattern, by default, we would only need Titan Text Embeddings V2  /  model id: _amazon.titan-embed-text-v2:0_.
+
+## Deployment Instructions
+
+1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
+    ``` 
+    git clone https://github.com/aws-samples/serverless-patterns
+    ```
+1. Change directory to the pattern directory:
```
[the remark this answers is not in the document]
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/README.md]: update sequence number

File: eventbridge-bedrock-s3-aoss/README.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+After you create your knowledge base, you ingest your data source/sources into your knowledge base so that they're indexed and are able to be queried. Additionally each time you add, modify, or remove files from your data source, you must sync the data source so that it is re-indexed to the knowledge base. Syncing is incremental, so Amazon Bedrock only processes added, modified, or deleted documents since the last sync.
+
+At the time of writing, Knowledge Bases for Amazon Bedrock doesn't have a feature to automatically periodically sync the datasource associated with a Knowledge Base. So customers who need to refresh their datasources periodically to ensure their knowledge base is up-to-date have to rely on bespoke solution. This pattern shows one way of implementing the solution, using [Amazon EventBridge Scheduler](https://docs.aws.amazon.com/scheduler/latest/UserGuide/what-is-scheduler.html).
```
[the remark this answers is not in the document]
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/README.md]: updated grammar

File: eventbridge-bedrock-s3-aoss/README.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+> This pattern uses Knowledge Bases for Amazon Bedrock and the Amazon Titan Text Embeddings V2. See [Supported regions and models for Knowledge bases for Amazon Bedrock](https://docs.aws.amazon.com/bedrock/latest/userguide/knowledge-base-supported.html) to select a region where Knowledge bases for Amazon Bedrock is supported
+
+## Enable Model Access on Amazon Bedrock
+Knowledge bases for Amazon Bedrock use a foundation model to embed your data sources in a vector store. Before creating a knowledge base and selecting an embeddings model for the Knowledge Base, You must request access to the model. If you try to use the model (with the API or console) before you have requested access to it, you receive an error message. For more information, see [Model access](https://docs.aws.amazon.com/bedrock/latest/userguide/model-access.html).
+
+1. In the AWS console, select the region from which you want to access Amazon Bedrock. 
+
+    ![Region Selection](images/region-selection.png)
+
+1. Find **Amazon Bedrock** by searching in the AWS console.
```
[the remark this answers is not in the document]
Comment (rajavaid77) [reply eventbridge-bedrock-s3-aoss/README.md]: updated

**Comment by rajavaid77** [comment]
> It seems that the pattern will periodically keep call Bedrock API to initiate KB sync even if there is no change in the source files? If so that will cause unnecessary API invocation costs. Instead of this, can we enhance this pattern to trigger the sync whenever there is a change in any source file and sync only that specific file from the bucket?



> I am getting this error while deploying:
> 
> `1:26:17 PM | CREATE_FAILED | AWS::Logs::Delivery | BedrockKBDelivery Resource handler returned message: "Supplied Policy document is breaching Cloudwatch Logs policy length limit. (Service: AWSLogs; Status Code: 400; Error Code: AccessDeniedException; Request ID: 5f1e6680-e223-4073-b440-033fadb60a0b; Proxy: null)" (RequestToken: e55167e6-3cf6-7017-b968-b5307db4d3e5, Handler ErrorCode: AccessDenied) 1:26:17 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack | BedrockKBStack The following resource(s) failed to create: [BedrockKBDelivery]. Rollback requested by user. 1:26:17 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack | BedrockKBStack The following resource(s) failed to create: [BedrockKBDelivery]. Rollback requested by user.`

fixed by updating the resource policy sourceARn

**Review by biswanathmukherjee — CHANGES_REQUESTED** [review]

**Comment by biswanathmukherjee** [comment]
Approved from my side. Thanks for your contribution!

**Pull request merged**